homeassistant/components/tplink_omada/update.py

```diff
@@ -97,53 +91,44 @@ def __init__(
-        self._attr_has_entity_name = True
-        self._attr_name = "Firmware Update"
-        self._refresh_state()
-
-    def _refresh_state(self) -> None:
-        if self._firmware_update and self._device.need_upgrade:
-            self._attr_installed_version = self._firmware_update.current_version
-            self._attr_latest_version = self._firmware_update.latest_version
-        else:
-            self._attr_installed_version = self._device.firmware_version
-            self._attr_latest_version = self._device.firmware_version
-        self._attr_in_progress = self._device.fw_download
-
-        if self._attr_in_progress:
-            # While firmware update is in progress, poll more frequently
-            async_call_later(self.hass, 60, self._request_refresh)
-
-    async def _request_refresh(self, _now: Any) -> None:
-        await self.coordinator.async_request_refresh()
 
     def release_notes(self) -> str | None:
         """Get the release notes for the latest update."""
-        if self._firmware_update:
-            return str(self._firmware_update.release_notes)
-        return ""
+        status = self.coordinator.data[self._mac]
+        if status.firmware:
+            return status.firmware.release_notes
+        return None
 
     async def async_install(
         self, version: str | None, backup: bool, **kwargs: Any
     ) -> None:
         """Install a firmware update."""
-        if self._firmware_update and (
-            version is None or self._firmware_update.latest_version == version
-        ):
-            await self._omada_client.start_firmware_upgrade(self._device)
-            await self.coordinator.async_request_refresh()
-        else:
-            _LOGGER.error("Firmware upgrade is not available for %s", self._device.name)
+        await self._omada_client.start_firmware_upgrade(
+            self.coordinator.data[self._mac].device
+        )
+        await self.coordinator.async_request_refresh()
 
     @callback
     def _handle_coordinator_update(self) -> None:
         """Handle updated data from the coordinator."""
         status = self.coordinator.data[self._mac]
-        self._device = status.device
-        self._firmware_update = status.firmware
-        self._refresh_state()
+
+        if status.firmware and status.device.need_upgrade:
+            self._attr_installed_version = status.firmware.current_version
+            self._attr_latest_version = status.firmware.latest_version
+        else:
+            self._attr_installed_version = status.device.firmware_version
+            self._attr_latest_version = status.device.firmware_version
+        self._attr_in_progress = status.device.fw_download
+
+        if self._attr_in_progress:
+            # While firmware update is in progress, poll more frequently
+            async def do_refresh(*_: Any) -> None:
+                await self.coordinator.async_request_refresh()
+
+            async_call_later(self.hass, 60, do_refresh)
```

Comment:
Should instead, the interval of the coordinator be adjusted?

Reply:
The problem is, I would have to set the refresh rate based on the firmware progress status of all of the devices that share the coodinator combined. I guess I could set the update interval from the polling function instead.

Reply:
>  I would have to set the refresh rate based on the firmware progress status of all of the devices

You are doing the same now. `async_request_refresh`  will trigger an update for all listeners.

Reply:
Yeah, it would have triggered an update callback for all `update` entities if one firmware _update_ was in progress. But in order to know what to set the poll interval to, the entity would need to check through all of the device statuses in the the coordinator data to see if all of them were not updating before setting the interval back to the default. It felt like the wrong place to be doing this.
I think the refactoring is a bit better, but there is more class boilerplate around each coordinator now.
Happy to revisit this, if you prefer.